[@types/object.groupby] Make output property types optional

huw opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
-[x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
-[x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
-[x] [Run pnpm test <package to test>](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [TC39 spec](https://tc39.es/proposal-array-grouping/)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

* * *

### Motivating Example

When running `Object.groupBy`, the specified callback function may not always output all of the possible types. When this happens, `Object.groupBy`, as per spec, will only include the keys that are actually output. To use an example, consider splitting an array into evens and odds ([TypeScript Playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMYdkAHAIQE8AeAFQBp4BpeEADwxFWAGd4AFYqRAwMlJiEoA+ABQAoeIvhZOMKACMIIAFzwAkqo1a6U+gqVgoECOqhgA1olS6ZANyvId8Bsu7tdqMgAtuoiAJTwALxSzGZhugBK4DgwwNRMjLQA2gC6UgDccnJgeDwY8HA8yBDlkYTEZFQyWQBMjAAsjABsOYxuHiAR0fDuEJ7wAKTwLVGRdQAM8AD88ABEIK5cq-C6qzjAwKthhZXVGAB0+8D5QA)):

```ts
const result = Object.groupBy([2, 4, 6], (value) => value % 2 === 0 ? "even" : "odd");
```

Here, `typeof result satisfies { even: number[]; odd: number[] }` but `JSON.stringify(result) === '{"even":[2,4,6]}'`. If we don’t know the value of the input array, it would be extremely unwise to start operating on `result.odd`.

### On `noUncheckedIndexedAccess`

For trivial examples, where `callbackFn` outputs `string` or `number` (ex. `Object.groupBy([1, 2, 3], value => value)`), we can solve this by enabling [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess), or we can opt out. This is presumably good, because it allows users to determine what level of type-safety they want.

However, `noUncheckedIndexedAccess` doesn’t apply to more complex types such as the even/odd example above. In that case, the user has no choice about the type-safety of their returns. In this example, with `noUncheckedIndexedAccess` enabled ([Playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMYdkAHAIQE8AeAFQBp4BpeEADwxFWAGd4AFYqRAwMlJiEoA+ABQAoeIvhZOMKACMIIAFzwAkqo1a6U+gqVgoECOqhgA1olS6ZANyvId8Bsu7tdqMgAtuoiAJTwALxSzGZhugBK4DgwwNRMjLQA2gC6UgDccnJgeDwY8HA8yBDlkYTEZFQyWQBMjAAsjABsOYxuHiAR0fDuEJ7wAKTwLVGRdQAM8AD88ABEIK5cq-C6qzjAwKthhSWoZfCc53WV1RgAdPvAhcWl5Tc1M3VEJBSUzQCMjDa8AAzL0RgMojFRp5ji8zuVLhhPhUQFUPll2jl8kA)):

```ts
const result = groupBy([2, 4, 6], (value) => value % 2 === 0 ? "even" : "odd");
const test = result.odd;

const result2 = groupBy([1, 2, 3], value => value);
const test2 = result2[4];
``` 

`typeof test2 satisfies number[] | undefined`, while `typeof test satisfies number[]`. This probably isn’t what users expect, or at least it’s not what I was expecting when I discovered this issue.

Specifically, the issue is that the unchecked index access rule only applies to unknown properties, but not mapped object types. When the output of `callbackFn` is a base type such as `string` or `number`, TypeScript treats these properties as ‘unknown’, since they’re infinite. But when `callbackFn` outputs a literal, or union of literals, these properties become ‘known’ and accessing them is no longer considered an unchecked index access.

### Potential Alternative
One alternative which would be somewhat less breaking would be to write (or similar):

```ts
declare function groupBy<T, K extends PropertyKey>(
    iterable: Iterable<T>,
    callbackfn: (value: T, index: number) => K,
): Record<K extends string ? string : K extends number ? number : K extends symbol ? symbol : never, T[]>;
```

This would unnarrow the key type, re-enabling the `noUncheckedIndexedAccess` behaviour, but losing the extra type information that would normally be inferable.

### Other Solutions Considered

I thought about all manner of clever ways to avoid having to always set every property to optional, but unfortunately there are no good solutions that feel simple enough to end up in `lib.d.ts`, which we’d want this to be eventually compatible with (if it doesn’t just directly adapt these types). There would be a lot of edge-casing more generally.

I specifically considered:
- Trying to determine if the output of `callbackFn` is a literal or array of literals, which would guarantee its availability (AFAIK there are no utility types which can distinguish `"even"` from `"even" | "odd"` and `string` which aren’t hacks)
- Guaranteeing that at least one possible option is required on the result type (using something like [`type-fest`’s `RequireAtLeastOne`](https://github.com/sindresorhus/type-fest/blob/a919f93384dba6e6d21e202a287986a6d4ea03e4/source/require-at-least-one.d.ts))

### Is this a breaking change?
I am not familiar with the versioning process for `@types` packages. This change will absolutely break the type-checking for existing code in existing repositories. On the other hand, it’s a bugfix which brings the types closer to the specification, and this package is a shim for a TC39 proposal so its types may not be expected to be stable? Any guidance here is appreciated and I’m happy to make those changes.

ljharb reviewed (changes requested):
I'm still pretty unclear on why this change is a good thing. That `result[4]` is undefined is indeed an unavoidable consequence of unchecked index access, and should have no impact on anything here.

huw commented:
Okay, I updated the original post to clarify! I was confused about the behaviour of `noUncheckedIndexedAccess`—you were right to note that the first example I gave with `groupBy([1, 2, 3], value => value)` is covered by that compiler flag. However, that only works for `groupBy`s with non-literal output types, and we’re not covered for literal output types; these won’t add `undefined` when accessing because those properties are considered ‘known’ by the compiler. I added a couple of playgrounds to the original post to illustrate what I mean.

ljharb commented:
I'm not sure how _useful_ this change is, though.

Certainly I hear you that not every potential key returned from the callback will be present on the object, although *at least one* will always be for an array of nonzero length - but certainly an empty array would create an empty object. However, I think this is just where TypeScript has decided to sit - not at "correct" or "sound", but at "useful".

huw commented:
I’m happy to defer to you in that case—it does sound like it’s verging into TypeScript’s territory and it would be better to raise an issue against their eventual implementation of it.

Thanks for your consideration!
